### ISSUE TITLE
Why silence model errors?

### ISSUE THREAD

**Issue by samedii** [Statisticat/LaplacesDemon#22]
Mo1 <- try(Model(prop, Data), silent=FALSE)

Is there a reason for this? Can it be set to TRUE instead or maybe an option for enabling it?

**Comment by samedii** [comment]
Also, why is the model inside a try?

**Comment by ipashchenko** [comment]
Good idea! I think user must know that something is wrong with his model. And ``try`` is enough (maybe sometimes user knows that problems but don't care for some reason)

**Comment by samedii** [comment]
@ipashchenko Could possibly keep the try and then if we have an error we can post that the model failed with these arguments. It would be hard to do more than that since the model function is specified by the user

**Comment by ipashchenko** [comment]
@samedii The same was once made in ``emcee`` python package - it made using it more comfortable. Such a internal debugging tool that shows where the problems are (and not only that the problems have happened)

**Comment by Statisticat** [comment]
Great idea. The "try" needs to be there, but it should be up to the user whether or not to see if the model was rejected because it failed. This could be put in a lot of functions in the package besides LaplacesDemon(). What's the best name for an argument? Debug=TRUE? Verbose? Any suggestions? Thanks for the good suggestion.

**Comment by samedii** [comment]
I kind of think it should be `silent=FALSE` by default and also stop running but if you are sure about what you're doing, you can force it to keep running. In command line apps such as grunt it's done with the flag "--force". So maybe `force=TRUE`?

Isn't it better if the user handles the errors instead of having the function crash?

**Comment by ipashchenko** [comment]
Well, ``DEBUG=TRUE`` is more explicitly i suppose... But ``silent`` is good too if this parameter is described in easily reachable places for users:)

**Comment by Statisticat** [comment]
I know one thing I wouldn't want to see happen, as a user, is update a complicated model for a long time, say a week, and then some weird exception happens, and it crashes and all that time is wasted. Agreed, I wouldn't want it to crash, but would rather the user knows there's a potential issue and they can decide whether or not to keep investing in time.

I'm going to look through the package now and think about silent=TRUE. That is, after all, exactly what the try code uses. If I were new to R and everything, and saw debug=TRUE, I'm guessing my first thought would be that some automatic debugging is going on, which it isn't. Hmm, I'm going to kick around "silent". I'm wondering where else it appears besides model evaluations, and if it should be explicit everywhere?

**Comment by samedii** [comment]
@Statisticat  Fair enough, sounds reasonable :+1:

**Comment by Statisticat** [comment]
Argh, if I were new to everything and saw silent=TRUE, my first fear would be there won't be any output at all. I guess clarity is an open issue, and then questions of where it should and should not be applied.

Just looking through LaplacesDemon(), try(...silent=TRUE) is used in a lot of places, such as for chol(), eigen(), FC() for Gibbs, MCSE, and more.

Any thoughts are welcome. I'm excited to have better debugging for users!

**Comment by Statisticat** [comment]
And if it inherits a try error, maybe we can insert more informative messages about where, when, and hopefully how!

**Comment by samedii** [comment]
Maybe a list?

So something like `showErrors=list(model=true, ...)` and if just `showErrors=TRUE` then all errors are shown? Some error messages may default to TRUE while others to FALSE this way

I'd would personally prefer seeing all crashing messages defaulting to being shown though

**Comment by ipashchenko** [comment]
OK, showing model errors by default sounds nice. But what about interrupting if model errors have occurred? I think there're two levels of options: 1) show messages on errors & 2) stop sampling (excecution) on erros

**Comment by samedii** [comment]
Stopping on errors isn't that important in my opinion. You will see the error in the log while it's running and so you can kill the demon manually if you want to

**Comment by Statisticat** [comment]
I'm also wondering about expandability to yet unforeseen debugging features.

Just kicking around ideas, but what do you think of something like:

debug=list(ShowErrors=list(chol=FALSE, eigen=FALSE, MCSE=FALSE, Model=TRUE, ...), ...)

A lot of cases for ShowErrors should default to FALSE in my opinion. For example, some adaptive algorithms try eigen(), if they fail they do something else, and there is nothing wrong, just an attempt by the algorithm.

I don't know what debugging options we would later include besides ShowErrors, but I'm sure something will get added later.

What do you think, or is there a better way of handling that?

**Comment by samedii** [comment]
Sounds good. If you can foresee more debug options in the near future then this seems like a good model. On the other hand it might not be needed while complicating syntax. But then I kind of like this progamming philosophy: [YAGNI](http://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) (in moderation anyway)

**Comment by Statisticat** [comment]
I've just added a Debug argument to LaplacesDemon and LaplacesDemon.hpc. Let me know what you think. When we're all happy with it, then I'll add it to the rest of the functions in the package.

**Comment by Statisticat** [comment]
Hopefully in a few hours I'll upload an expansion on the Debug feature. I'll go sampler by sampler and put in informative error messages.

A neat feature is that in some samplers we can learn a lot. For example, if it is a componentwise sampler, it can report which parameter was changed, and if a non-finite value resulted, or what not.

If it is slice, we can also know if it occurred when testing a lower or upper interval bound.

If it is HMC, we can know if it happened in the leapfrog (most likely), or in the final acceptance step.

On and on.

Richard, I'm also incorporating your nested "else if" for the acceptance step, because that should speed things up a smidgeon, and I take speed of an R-based sampler seriously.

None of these debugging messages should slow it down, because they only occur when it's goofing.

Anyway, I'm anxious to see what is thought of this next upload.

**Comment by samedii** [comment]
@Statisticat Sounds like a great addition! With risk of saying something you already though about, some general default of just showing when and with what values something failed could be useful when the code doesn't catch exactly what went wrong (and letting the user figure it out)

Gonna check it out!

**Comment by Statisticat** [comment]
I'll post an update soon that gives even more detail to componentwise sampler warnings. Currently, it lets you know which parameter name went goofy, and soon it will also report the iteration, the current value, and the proposed value. It also currently reports some conditions within the sampler. For example, if using slice sampling, it will tell you if it failed trying to step out of the lower bound, upper bound, or during rejection sampling. For Hamiltonian methods, it will tell you which leapfrog step it happened it. This is neat, but...

Multivariate samplers on the other hand, I'm scratching my head wondering what to report. The iteration should be reported of course, but what about parameters? There may not be a good answer. For example, a non-finite LP may occur, but not due to any non-finite proposed parameter. I suppose it should check for non-finite proposals and report them, but it probably shouldn't print all parameters to the screen...someone running a model with 10,000 parameters would flood their buffer quickly with a few warnings...and there's no meaningful way to look through it for a problem.

Any suggestions are welcome of course.

**Comment by samedii** [comment]
Could maybe log long errors in a file and print a summary to the console including the location of this log file?

**Comment by Statisticat** [comment]
I think I have it, what do you think:

For a multivariate sampler, it reports the iteration and proposal. Even though the proposal could be huge, the user always has the option of using the LogFile argument to redirect output like that elsewhere. But, assuming this huge proposal vector fills your screen, it will be converted via paste and collapse to look like c(123,345,...), so that if you want to, you can stop the stampler, copy/paste the proposal, run it through your model function, and investigate it on your own.

**Comment by samedii** [comment]
Sounds good :+1:

**Comment by Statisticat** [comment]
That's been uploaded for multivariate sampler reporting.
